src/Sentry/Internal/CachingBackgroundWorker.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol.Envelopes;
+
+namespace Sentry.Internal
+{
+    internal class CachingBackgroundWorker : BackgroundWorkerBase
+    {
+        private readonly DirectoryInfo _cacheDirectory;
+
+        public CachingBackgroundWorker(ITransport transport, SentryOptions options)
+            : base(transport, options)
+        {
+            _cacheDirectory =
+                options.CacheDirectory ??
+                throw new InvalidOperationException("Cache directory is not set.");
+        }
+
+        private IEnumerable<FileInfo> GetCachedEnvelopeFiles()
+        {
+            if (!_cacheDirectory.Exists)
+            {
+                return Enumerable.Empty<FileInfo>();
+            }
+
+            return _cacheDirectory.EnumerateFiles("*.envelope");
+        }
+
+        private FileInfo? TryGetNextEnvelopeFile() => GetCachedEnvelopeFiles()
+            .OrderBy(f => f.CreationTimeUtc)
+            .FirstOrDefault();
+
+        private async ValueTask FlushCacheAsync(CancellationToken cancellationToken = default)
+        {
+            while (TryGetNextEnvelopeFile() is { } envelopeFile)
+            {
+                Options.DiagnosticLogger?.LogDebug("Sending cached envelope: {0}", envelopeFile.FullName);
+
+                using var envelope = await Envelope.DeserializeAsync(
+                    envelopeFile.OpenRead(),
+                    cancellationToken
+                ).ConfigureAwait(false);
+
+                await Transport.SendEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
+
+                envelopeFile.Delete();
+            }
+        }
+
+        private async ValueTask CacheEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken = default)
+        {
+            if (GetCachedEnvelopeFiles().Count() >= 50)
```

Comment:
30 I believe is the general default, this will need to propagate from Options.

```suggestion
            if (GetCachedEnvelopeFiles().Count() >= 30)
```

Reply:
Yeah, I just put a random number for now